File: server/Game.cpp
#include "Game.h"
#include "gamestates/PlayingState.h"
#include "exceptions/InvalidActionException.h"
#include "exceptions/InvalidMoveException.h"
#include "exceptions/NoSuchPieceException.h"
#include "exceptions/NotYourTurnException.h"
#include "exceptions/FirstTurnException.h"
#include "Player.h"
#include "Move.h"
#include "PawnPromotionStrategy.h"
#include "ChessMatrix.h"

Game::Game (int gameId, Player* playerW, Player* playerB): 
    _gameId (gameId), _playerW(playerW), _playerB(playerB), _lastMove(nullptr)
{
    _whiteTurn = false;
    _chessMatrix = new ChessMatrix;

    for(int x = 0; x < MAX_X; x++)
    {
        for(int y = 0; y < MAX_Y; y++)
        {
            _chessMatrix->set(Position(x,y), nullptr);
        }
    }
    
    boardCreation();
    createDrawConditions(); 

    _state = new PlayingState(this);
}

Game::~Game ()
{
    for (Piece* piece: _promoted)
        delete piece;
    
    delete _state;
    delete _chessMatrix;
}

void Game::createDrawConditions()
{ 
    /*std::list<Piece*> condition;
    
    _drawCondition.push_front();
    _drawCondition.push_front(_bishopB[0]);
    _drawCondition.push_front(_bishopB[1]);
    _drawCondition.push_front(_bishopW[0]);
    _drawCondition.push_front(_bishopW[1]);

    _drawCondition.push_front(_knightB[0]);
    _drawCondition.push_front(_knightB[1]);
    _drawCondition.push_front(_knightW[0]);
    _drawCondition.push_front(_knightW[1]);*/
}

void Game::boardCreation()
{
    bool forward = true;

    int firstYB = 6;
    int secondYB = 7;
    int firstYW = 1;
    int secondYW = 0;
    
    Position posW(0,firstYW);
    Position posB(0,firstYB);

    for(int x = 0; x < MAX_X; x++)
    {
        posW.x = x; 
        posB.x = x;

        _pawnW.push_back(PawnPiece(x*2, true, posW, this, forward) );
        _pawnB.push_back(PawnPiece(x*2+1, false, posB, this, !forward) );
        
        _chessMatrix->set(posB, &(_pawnB.back()) );
        _chessMatrix->set(posW, &( _pawnW.back()) );
    }

    posW.y = secondYW;
    posB.y = secondYB;

    int rookId = 0, knightId = 0, queenId = 0, bishopId = 0, kingId = 0;
    for(int x = 0; x < MAX_X; x++)
    {
        posW.x = x;
        posB.x = x; 
        if(x==0 || x == MAX_X - 1)
        {
            _rookW.push_back(RookPiece(rookId*2, true, posW, this, forward) );
            _rookB.push_back(RookPiece(rookId*2+1, false, posB, this, !forward) );
            _chessMatrix->set(posB, &_rookB.back());
            _chessMatrix->set(posW, &_rookW.back());
            rookId++;
        }
        else if(x==1 || x == MAX_X -2)
        {
            _knightW.push_back(KnightPiece(knightId*2, true, posW, this, forward) );
            _knightB.push_back(KnightPiece(knightId*2+1, false, posB, this, !forward) );
            _chessMatrix->set(posB, &_knightB.back());
            _chessMatrix->set(posW, &_knightW.back());
            knightId++;
        }
        else if(x==2 || x == MAX_X -3)
        {
            _bishopW.push_back(BishopPiece(bishopId*2, true, posW, this, forward) );
            _bishopB.push_back(BishopPiece(bishopId*2+1, false, posB, this, !forward) );
            _chessMatrix->set(posB, &_bishopB.back());
            _chessMatrix->set(posW, &_bishopW.back());
            bishopId++;
        }
        else if(x==3)
        {
            _kingW.set(kingId*2, true, posW, this, forward);
            _kingB.set(kingId*2+1, false, posB, this, !forward);
            _chessMatrix->set(posB, &_kingB);
            _chessMatrix->set(posW, &_kingW);
            kingId++;
        }  
        else if(x==4)
        {
            _queenW.push_back(QueenPiece(queenId*2, true, posW, this, forward) );
            _queenB.push_back(QueenPiece(queenId*2+1, false, posB, this, !forward) );
            _chessMatrix->set(posB, &_queenB.back());
            _chessMatrix->set(posW, &_queenW.back());
            queenId++;
        }
    }
}

void Game::move(const Position& origin, const Position& dest,
            const std::string& userId) 
{
    _state->move(origin, dest, userId);
    setLastMove(Move(origin, dest));
}
 
void Game::fillEnPassant(const Position& lastPos,Piece* piece)
{
    bool white = piece->isWhite();
    std::list<Position> posList;
    Position& pos = piece->getPos();
    if(!white)
    {
        for(PawnPiece& p : _pawnW)
        {
            if(p.getPos().y == pos.y && (p.getPos().x+1 == 
                pos.x || p.getPos().x-1 == pos.x) )
            {
                posList.push_front(p.getPos());
            }
        }
    }
    else
    {
        for(PawnPiece& p : _pawnB)
        {
            if(p.getPos().y == pos.y && (p.getPos().x+1 == 
                pos.x || p.getPos().x-1 == pos.x) )
            {
                posList.push_front(p.getPos());
            }
        }        
    }

    _chessMatrix->setEnPassant(piece, posList,lastPos);
}

void Game::removePawn(PawnPiece* p)
{
    if(p->isWhite())
    {
        for(auto it = _pawnW.begin(); it != _pawnW.end(); it++)
        {
            if((*it).getId()==p->getId())
            {
                _pawnW.erase(it);
            }
        }
    }
    else
    {
        for(auto it = _pawnB.begin(); it != _pawnB.end(); it++)
        {
            if((*it).getId()==p->getId())
            {
                _pawnB.erase(it);
            }
        }
    }
}

const Move& Game::lastMove ()
{
    if (_lastMove == nullptr) 
        throw FirstTurnException ();
    else
        return *_lastMove;
}

void Game::setLastMove (const Move& move)
{
    if (_lastMove != nullptr)
        delete _lastMove;
    
    _lastMove = new Move (move.origin, move.destination);
}

void Game::promote(PawnPromotionStrategy* strategy)
{
    //Inserting the piece must be done by the messages otherwise
    // it needs 6 different methods or a instanceof
    _state->promote(strategy);
}

bool Game::drop (const std::string& userId)
{
    _state->drop(userId);
}

const std::string& Game::getTurnUser () 
{
    return _whiteTurn ? _playerW->user() : _playerB->user();
}

void Game::tickTurn() 
{
    _whiteTurn = !_whiteTurn; 
    _chessMatrix->tickTurn(); 
}

std::list<Position>& Game::getEnPassantOrigin() 
{ 
    return _chessMatrix->getEnPassantOrigin();
}

Position* Game::getEnPassantDest() 
{ 
    return _chessMatrix->getEnPassantDest();
}

Piece* Game::getEnPassantPiece() 
{
    return _chessMatrix->getEnPassantPiece();
}

int Game::getEnPassantLiveTime() {
    return _chessMatrix->getEnPassantLiveTime();
}

Piece* Game::getCell(const Position& pos) 
{
    return _chessMatrix->get(pos);
}

#include <iostream>
/*
//Removing this will remove the text simulation
int main()
{
    Game game(0,nullptr,nullptr); 
    game.printMatrix();
    //ChessMatrix* m = game->getMatrix();
    m->printMatrix();//
    int xOld,xNew,yOld,yNew;
    bool promote = false;
    while(true)
    {
        std::cin >> xOld >> yOld >> xNew >> yNew;

        Position lastPos(xOld,yOld);
        Position newPos(xNew,yNew);
        try
        {
            if(promote==false)
                game.move(lastPos,newPos); 
            else
            {
                QueenPiece q = QueenPiece();
                game.promote(&q);
                game.insertQueen(q.isWhite(),q);
                promote = false;
            }
        }catch(PawnPromotionException& e)
        {
            promote = true;
        }catch( const std::exception& e)
        {
            std::cout<< "Error: " << e.what() <<std::endl;
        }

        std::cout << std::endl << std::endl;
        
        game.printMatrix(); 
    }
    return 0;
}*/
/*
1 7 0 5
0 1 0 2
0 6 0 5
1 6 1 5
0 2 0 3
2 7 1 6
0 3 0 4
*/

File: client/js/Game.js
class Game {
    constructor(gameId, newGame, whiteTurn, currentState) {
        this.gameId = gameId;
        if(this.whiteTurn == undefined) {

            this.whiteTurn = true;
        }
        else {
            this.whiteTurn = whiteTurn;
        }
        this.chessMatrix = new Board();

        this.kingW = null; this.kingB = null;
        this.queenW = []; this.queenB = [];
        this.pawnW = []; this.pawnB = [];
        this.rookW = []; this.rookB = [];
        this.knightW = []; this.knightB = [];
        this.bishopW = []; this.bishopB = [];

        for(var x = 0; x < BOARD_MAX_X; x++)
        {
            for(var y = 0; y < BOARD_MAX_Y; y++)
            {
                this.chessMatrix.set(new Position(x,y), null);
            }
        }

        if(currentState == undefined) {
            this.state = new PlayingState(this);
        }
        else {
            this.state = currentState;
        }
        
        if(newGame) {
            this.createBoardFromNothing();
        }

        this.meWhite= true//FIXME
    }

    //IDEA FOR IMPORTING AN EXISTING GAME FROM THE SERVER
    //fullBoardImport()
    createBoardFromNothing() {
        var forward = true;
        var firstYB = 6;
        var secondYB = 7;
        var firstYW = 1;
        var secondYW = 0;

        var posW = new Position(0, firstYW);
        var posB = new Position(0, firstYB);


        for(var x = 0; x < BOARD_MAX_X; x++) {
            posW.x = x;
            posB.x = x;

            this.pawnW.push(new PawnPiece(x*2, true, posW, this, forward));
            this.pawnB.push(new PawnPiece(x*2+1, false, posB, this, !forward));

            this.chessMatrix.set(posB, this.pawnB[this.pawnB.length - 1]);
            this.chessMatrix.set(posW, this.pawnW[this.pawnW.length - 1]);
        }

        posW.y = secondYW;
        posB.y = secondYB;
        var rookId = 0, knightId = 0, queenId = 0, bishopId = 0, kingId = 0;
        for(var x = 0; x < BOARD_MAX_X; x++) {
            posW.x = x;
            posB.x = x;
            if(x==0 || x==BOARD_MAX_X - 1) {

                this.rookW.push(new RookPiece(rookId*2, true, posW, this, forward));
                this.rookB.push(new RookPiece(rookId*2+1, false, posB, this, !forward));
    
                this.chessMatrix.set(posB, this.rookB[this.rookB.length - 1]);
                this.chessMatrix.set(posW, this.rookW[this.rookW.length - 1]);
                rookId++;
            }
            else if(x==1 || x==BOARD_MAX_X - 2) {

                this.knightW.push(new KnightPiece(knightId*2, true, posW, this, forward));
                this.knightB.push(new KnightPiece(knightId*2+1, false, posB, this, !forward));
    
                this.chessMatrix.set(posB, this.knightB[this.knightB.length - 1]);
                this.chessMatrix.set(posW, this.knightW[this.knightW.length - 1]);
                knightId++;
            }
            else if(x==2 || x==BOARD_MAX_X - 3) {

                this.bishopW.push(new BishopPiece(bishopId*2, true, posW, this, forward));
                this.bishopB.push(new BishopPiece(bishopId*2+1, false, posB, this, !forward));
    
                this.chessMatrix.set(posB, this.bishopB[this.bishopB.length - 1]);
                this.chessMatrix.set(posW, this.bishopW[this.bishopW.length - 1]);
                bishopId++;this
            }
            else if(x==3) {

                this.kingW = new KingPiece(kingId*2, true, posW, this, forward);
                this.kingB = new KingPiece(kingId*2+1, false, posB, this, !forward);

                this.chessMatrix.set(posB,this.kingB);
                this.chessMatrix.set(posW,this.kingW);
            }
            else if(x==4) {

                this.queenW.push(new QueenPiece(queenId*2, true, posW, this, forward));
                this.queenB.push(new QueenPiece(queenId*2+1, false, posB, this, !forward));
    
                this.chessMatrix.set(posB, this.queenB[this.queenB.length - 1]);
                this.chessMatrix.set(posW, this.queenW[this.queenW.length - 1]);
                queenId++;
            }
        }

    }



    checkMove(origin,dest) {
        return this.state.checkMove(origin, dest);
    }
    move(origin, dest) {
        this.state.move(origin,dest);
    }

    getGameId() { return this.gameId; }

    //FIXME PROMOTE DO I CREATE THE METHOD PROMOTE OR DO WE JUST SET THE POSITIONS
    //FROM THE REQUEST
    promote(piece) {
        
    }

    getCell(pos) { return this.chessMatrix.get(pos); }

    tickTurn() {
        this.whiteTurn = !this.whiteTurn;
        this.chessMatrix.tickTurn();
    }

    fillEnPassant(lastPos, piece) {
        let white = piece.white;
        let posList = [];
        let pos = piece.pos;
        if(!white) {
            for(let i in this.pawnW) {
                let p = this.pawnW[i];
                if(p.pos.y == pos.y && (p.pos.x + 1 == pos.x ||
                    p.pos.x-1 == pos.x) ) {
                    
                    posList.push(p.pos);
                }
            }
        }
        else {
            for(let i in this.pawnB) {
                let p = this.pawnB[i];
                if(p.pos.y == pos.y && (p.pos.x + 1 == pos.x ||
                    p.pos.x-1 == pos.x) ) {
                    
                    posList.push(p.pos);
                }
            }
        }

        this.chessMatrix.setEnPassant(piece, posList, lastPos);
    }

    getEnPassantOrigin() { return this.chessMatrix.getEnPassantOrigin(); }
    getEnPassantDest() { return this.chessMatrix.getEnPassantDest(); }
    getEnPassantPiece() { return this.chessMatrix.getEnPassantPiece(); }

    getBishop(white) {
        if(white) {return this.bishopW;} else {return this.bishopB;}
    } 
    getKnight(white) {
        if(white) {return this.knightW;} else {return this.knightB;}
    }
    getRook(white) {
        if(white) {return this.rookW;} else {return this.rookB;}
    }
    getPawn(white) {
        if(white) {return this.pawnW;} else {return this.pawnB;}
    }
    getQueen(white) {
        if(white) {return this.queenW;} else {return this.queenB;}
    }
    getKing(white) {
        if(white) {return this.kingW;} else {return this.kingB;}
    }
    //setState ?

    removePawn(p) {
        if(p.white) {
            for(let i = 0; i < this.pawnW.length; i++) {
                if(this.pawnW[i].id == p.id) {
                    this.pawnW.splice(i,1);
                }
            }
        }
        else {
            for(let i = 0; i < this.pawnB.length; i++) {
                if(this.pawnB[i].id == p.id) {
                    this.pawnB.splice(i,1);
                }
            }
        }
    }
 
    insertBishop(white, p) {
        if(white) {this.bishopW.push(p);} else {this.bishopB.push(p);}
    } 
    insertKnight(white, p) {
        if(white) { this.knightW.push(p);} else { this.knightB.push(p);}
    }
    insertRook(white, p) {
        if(white) { this.rookW.push(p);} else { this.rookB.push(p);}
    }
    insertPawn(white, p) {
        if(white) { this.pawnW.push(p);} else { this.pawnB.push(p);}
    }
    insertQueen(white, p) {
        if(white) { this.queenW.push(p);} else { this.queenB.push(p);}
    }

    /////////////////////////////////VISUAL
    translatePosIntoVisual(position) {
        return this.chessMatrix.translatePosIntoVisual(position);
    }

    getVisualPieces() {
        var visual = [];
        //MISSING VISUALS!
        for(var i in this.pawnB)
            visual.push(this.pawnB[i].visual);
        for(var i in this.pawnW)
            visual.push(this.pawnW[i].visual);

        for(var i in this.queenB)
            visual.push(this.queenB[i].visual);
        for(var i in this.queenW)
            visual.push(this.queenW[i].visual);

        for(var i in this.knightB)
            visual.push(this.knightB[i].visual);
        for(var i in this.knightW)
            visual.push(this.knightW[i].visual);

        for(var i in this.bishopB)
            visual.push(this.bishopB[i].visual);
        for(var i in this.bishopW)
            visual.push(this.bishopW[i].visual);

        for(var i in this.rookB)
            visual.push(this.rookB[i].visual);
        for(var i in this.rookW)
            visual.push(this.rookW[i].visual);
        
        visual.push(this.kingB.visual);
        visual.push(this.kingW.visual);
        
        return visual;
    }

    getBoardVisualTiles() {
        return this.chessMatrix.getBoardVisualTiles();
    }

    update(deltaTime) {
        this.chessMatrix.update();
    }
    //DROP FIXME
    // drop(white) { }
    
}